oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/JwtIssuerAuthenticationManagerResolver.java

```diff
@@ -167,15 +170,25 @@ public String convert(@NonNull BearerTokenAuthenticationToken authentication) {
 
 	}
 
-	static class TrustedIssuerJwtAuthenticationManagerResolver implements AuthenticationManagerResolver<String> {
+	/**
+	 * An implementation of {@link AuthenticationManagerResolver} that validates the JWT issuer before
+	 * resolving the {@link AuthenticationManager} for the issuer.
+	 */
+	public static class TrustedIssuerJwtAuthenticationManagerResolver
+			implements AuthenticationManagerResolver<String> {
 
 		private final Log logger = LogFactory.getLog(getClass());
 
 		private final Map<String, AuthenticationManager> authenticationManagers = new ConcurrentHashMap<>();
 
 		private final Predicate<String> trustedIssuer;
 
-		TrustedIssuerJwtAuthenticationManagerResolver(Predicate<String> trustedIssuer) {
+		/**
+		 * Constructs a {@link TrustedIssuerJwtAuthenticationManagerResolver} instance using the predicate
+		 * @param trustedIssuer a predicate to validate issuer
+		 */
+		public TrustedIssuerJwtAuthenticationManagerResolver(Predicate<String> trustedIssuer) {
```

Comment:
I prefer the other idea that you listed of adding it as a top-level constructor. That allows this class to remain private for now.

Reply:
What about instead deprecating the trusted-issuer constructors and changing to static factory methods? I think this is nice in that it addressed the backward compatibility concern as well as aligns more closely with related components like `NimbusJwtDecoder`.

I think the method `fromTrustedIssuers` makes sense:

```
AuthenticationManager#fromTrustedIssuers(String...)
AuthenticationManager#fromTrustedIssuers(Collection<String>)
AuthenticationManager#fromTrustedIssuers(Predicate<String>)
```

Reply:
That's a great idea! I wasn't sure there was an appetite for such a change. Do you want to cherry-pick the other 3 commits? I'll refactor the last commit with the proposed static factory methods.

Reply:
@jzheaux I have made the requested changes. Please have another look.